### PR TITLE
libpulsar: 2.10.2 -> 3.5.0, pythonPackages.pulsar: init at 3.4.1

### PR DESCRIPTION
**Pull request by gaelreyrol**
###### Description of changes

This PR updates libpulsar to its latest [version](https://github.com/apache/pulsar-client-cpp/releases/tag/v3.3.0) and adds a new Python package to support the standalone Python [repository](https://github.com/apache/pulsar-client-python).

Python wrapper inside libpulsar has been removed from main package `libpulsar`.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/doc/manual/release-notes/rl-2311.section.md [hunk 1]
```diff
@@ -80,6 +80,8 @@
 
 - The Caddy module gained a new option named `services.caddy.enableReload` which is enabled by default. It allows reloading the service instead of restarting it, if only a config file has changed. This option must be disabled if you have turned off the [Caddy admin API](https://caddyserver.com/docs/caddyfile/options#admin). If you keep this option enabled, you should consider setting [`grace_period`](https://caddyserver.com/docs/caddyfile/options#grace-period) to a non-infinite value to prevent Caddy from delaying the reload indefinitely.
 
+- `libpulsar` Python wrapper support has been removed and is now a standalone repository. See [repository](https://github.com/apache/pulsar-client-python) for more details.
```
Comment (SuperSandro2000) [inline nixos/doc/manual/release-notes/rl-2311.section.md:83]: We should mention where to find that in nixpkgs

File: pkgs/development/python-modules/pulsar/default.nix [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  ];
+
+  propagatedBuildInputs = [
+    certifi
+  ];
+
+  # Requires to setup a cluster
+  doCheck = false;
+
+  passthru.libpulsar-python = libpulsar-python;
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/pulsar/default.nix:59]:
```suggestion
  passthru = {
    inherit libpulsar-python;
  };
```

File: pkgs/development/python-modules/pulsar/default.nix [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+  };
+in buildPythonPackage {
+  pname = "pulsar";
+  inherit src version;
+  format = "setuptools";
+
+  disabled = pythonOlder "3.7";
+
+  preConfigure = ''
+    cp ${libpulsar-python}/lib/lib_pulsar.so .
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/pulsar/default.nix:45]: Can we symlink this instead?

File: pkgs/development/python-modules/pulsar/default.nix [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+let
+  version = "3.2.0";
+  src = fetchFromGitHub {
+    owner = "apache";
+    repo = "pulsar-client-python";
+    rev = "v${version}";
+    hash = "sha256-7YFU7l4NKEwMPWo+XVKfX40eFun+AZBDStaT6KeKJSY=";
+  };
+  libpulsar-python = stdenv.mkDerivation {
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/pulsar/default.nix:22]: We can't do this in one build?

File: pkgs/development/python-modules/pulsar/default.nix [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    pname = "libpulsar-python";
+    inherit version src;
+
+    nativeBuildInputs = [ cmake pkg-config ];
+
+    buildInputs = [ libpulsar python pybind11 ];
+
+    enableParallelBuilding = true;
+
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/pulsar/default.nix:31]:
```suggestion
```
already done by cmake hook

File: pkgs/development/python-modules/pulsar/default.nix [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    postInstall = ''
+      mkdir -p $out/lib
+      cp lib_pulsar.so $out/lib
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/pulsar/default.nix:34]: This won't work on darwin

File: nixos/doc/manual/release-notes/rl-2311.section.md [hunk 7]
```diff
@@ -88,6 +88,8 @@
 
 - The `go-ethereum` package has been updated to v1.12.0. This drops support for proof-of-work. Its GraphQL API now encodes all numeric values as hex strings and the GraphQL UI is updated to version 2.0. The default database has changed from `leveldb` to `pebble` but `leveldb` can be forced with the --db.engine=leveldb flag. The `checkpoint-admin` command was [removed along with trusted checkpoints](https://github.com/ethereum/go-ethereum/pull/27147).
 
+- `libpulsar` Python wrapper support has been removed and is now a standalone Python package `pythonPackages.pulsar`. See new standalone [repository](https://github.com/apache/pulsar-client-python) for more details.
```
Comment (SuperSandro2000) [inline nixos/doc/manual/release-notes/rl-2311.section.md:91]: It should be enough to add an alias for this or are there any other changes that people must know?

File: pkgs/development/python-modules/pulsar/default.nix [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    owner = "apache";
+    repo = "pulsar-client-python";
+    rev = "v${version}";
+    hash = "sha256-7YFU7l4NKEwMPWo+XVKfX40eFun+AZBDStaT6KeKJSY=";
+  };
+
+  disabled = pythonOlder "3.7";
+
+
+  nativeBuildInputs = [
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/pulsar/default.nix:27]:
```suggestion
  disabled = pythonOlder "3.7";

  nativeBuildInputs = [
```

File: nixos/doc/manual/release-notes/rl-2311.section.md (hunk 7 above)
Comment (gaelreyrol) [reply nixos/doc/manual/release-notes/rl-2311.section.md:91]: It should be enough.

Comment (SuperSandro2000) [reply nixos/doc/manual/release-notes/rl-2311.section.md]: So we can drop the entire changelog entry and add the alias or why not?

**Comment by CorbanR** [comment]
I run OSX with an M2 processor. 

I ran `nix-build -E 'with import <nixpkgs> { }; callPackage ./default.nix { }'` on `pkgs/development/libraries/libpulsar/` and it works just fine on OSX

But I can't seem to get the python-module to build `pkgs/development/python-modules/pulsar/`
The command I tried was `nix-build -E 'with import <nixpkgs> {}; pkgs.python3Packages.callPackage ./default.nix {}'`  

 The closest I got was using `python38Packages` but maybe I'm running the wrong command to test with. Any suggestions?

**Comment by gaelreyrol** [comment]
@CorbanR Thanks for the feedback, unfortunately I am not able to troubleshoot anything because I don't have an Apple device. What errors did you get, It might be interesting to post them here so OSX users can trace that it was initially not working ?

**Comment by gaelreyrol** [comment]
@CorbanR Can you try again ? A new release for Python has been published and it seems they fixed an OpenSSL errors on MacOS. https://github.com/apache/pulsar-client-python/releases/tag/v3.3.0

File: nixos/doc/manual/release-notes/rl-2311.section.md (hunk 7 above)
Comment (SuperSandro2000) [reply nixos/doc/manual/release-notes/rl-2311.section.md]: please see https://github.com/NixOS/nixpkgs/blob/master/pkgs/top-level/aliases.nix

Comment (gaelreyrol) [reply nixos/doc/manual/release-notes/rl-2311.section.md]: Oh I see but how can I restrict it in usage of a "Python" derivation?

**Comment by CorbanR** [comment]
@gaelreyrol I have retried and it works! Good work!

**Comment by CorbanR** [comment]
I ran into an error building with python312, but I don't think that should hold this up. Building with python310, and python311 works fine. 

python312 error
```
Building setuptools wheel...
/private/tmp/nix-build-python3.12-bootstrapped-pip-23.0.1.drv-0/setuptools /private/tmp/nix-build-python3.12-bootstrapped-pip-23.0.1.drv-0
ERROR: Exception:
Traceback (most recent call last):
  File "/private/tmp/nix-build-python3.12-bootstrapped-pip-23.0.1.drv-0/pip/src/pip/_internal/cli/base_command.py", line 160, in exc_logging_wrapper
    status = run_func(*args)
             ^^^^^^^^^^^^^^^
  File "/private/tmp/nix-build-python3.12-bootstrapped-pip-23.0.1.drv-0/pip/src/pip/_internal/cli/req_command.py", line 247, in wrapper
    return func(self, options, args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/private/tmp/nix-build-python3.12-bootstrapped-pip-23.0.1.drv-0/pip/src/pip/_internal/commands/install.py", line 345, in run
    session = self.get_default_session(options)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/private/tmp/nix-build-python3.12-bootstrapped-pip-23.0.1.drv-0/pip/src/pip/_internal/cli/req_command.py", line 98, in get_default_session
    self._session = self.enter_context(self._build_session(options))
                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/private/tmp/nix-build-python3.12-bootstrapped-pip-23.0.1.drv-0/pip/src/pip/_internal/cli/req_command.py", line 125, in _build_session
    session = PipSession(
              ^^^^^^^^^^^
  File "/private/tmp/nix-build-python3.12-bootstrapped-pip-23.0.1.drv-0/pip/src/pip/_internal/network/session.py", line 343, in __init__
    self.headers["User-Agent"] = user_agent()
                                 ^^^^^^^^^^^^
  File "/private/tmp/nix-build-python3.12-bootstrapped-pip-23.0.1.drv-0/pip/src/pip/_internal/network/session.py", line 175, in user_agent
    setuptools_dist = get_default_environment().get_distribution("setuptools")
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/private/tmp/nix-build-python3.12-bootstrapped-pip-23.0.1.drv-0/pip/src/pip/_internal/metadata/importlib/_envs.py", line 188, in get_distribution
    return next(matches, None)
           ^^^^^^^^^^^^^^^^^^^
  File "/private/tmp/nix-build-python3.12-bootstrapped-pip-23.0.1.drv-0/pip/src/pip/_internal/metadata/importlib/_envs.py", line 183, in <genexpr>
    matches = (
              ^
  File "/private/tmp/nix-build-python3.12-bootstrapped-pip-23.0.1.drv-0/pip/src/pip/_internal/metadata/base.py", line 612, in iter_all_distributions
    for dist in self._iter_distributions():
  File "/private/tmp/nix-build-python3.12-bootstrapped-pip-23.0.1.drv-0/pip/src/pip/_internal/metadata/importlib/_envs.py", line 176, in _iter_distributions
    for dist in finder.find_eggs(location):
  File "/private/tmp/nix-build-python3.12-bootstrapped-pip-23.0.1.drv-0/pip/src/pip/_internal/metadata/importlib/_envs.py", line 144, in find_eggs
    yield from self._find_eggs_in_dir(location)
  File "/private/tmp/nix-build-python3.12-bootstrapped-pip-23.0.1.drv-0/pip/src/pip/_internal/metadata/importlib/_envs.py", line 111, in _find_eggs_in_dir
    from pip._vendor.pkg_resources import find_distributions
  File "/private/tmp/nix-build-python3.12-bootstrapped-pip-23.0.1.drv-0/pip/src/pip/_vendor/pkg_resources/__init__.py", line 2164, in <module>
    register_finder(pkgutil.ImpImporter, find_on_path)
                    ^^^^^^^^^^^^^^^^^^^
AttributeError: module 'pkgutil' has no attribute 'ImpImporter'. Did you mean: 'zipimporter'?

nix-build -E 'with import <nixpkgs> { }; callPackage ./default.nix { }'error: builder for '/nix/store/nrz69savf1nysqz6skrmp2hwsab5zc2q-python3.12-bootstrapped-pip-23.0.1.drv' failed with exit code 2;
       last 10 log lines:
       >     for dist in finder.find_eggs(location):
       >   File "/private/tmp/nix-build-python3.12-bootstrapped-pip-23.0.1.drv-0/pip/src/pip/_internal/metadata/importlib/_envs.py", line 144, in find_eggs
       >     yield from self._find_eggs_in_dir(location)
       >   File "/private/tmp/nix-build-python3.12-bootstrapped-pip-23.0.1.drv-0/pip/src/pip/_internal/metadata/importlib/_envs.py", line 111, in _find_eggs_in_dir
       >     from pip._vendor.pkg_resources import find_distributions
       >   File "/private/tmp/nix-build-python3.12-bootstrapped-pip-23.0.1.drv-0/pip/src/pip/_vendor/pkg_resources/__init__.py", line 2164, in <module>
       >     register_finder(pkgutil.ImpImporter, find_on_path)
       >                     ^^^^^^^^^^^^^^^^^^^
       > AttributeError: module 'pkgutil' has no attribute 'ImpImporter'. Did you mean: 'zipimporter'?
       >
```

**Review by CorbanR — APPROVED** [review]

**Comment by gaelreyrol** [comment]
@CorbanR the main repository appears to be tested on Python 11, not higher so I guess it is "normal"

**Comment by SuperSandro2000** [comment]
> @CorbanR the python client repository appears to be tested on Python 11, not higher so I guess it is "normal". Should I prevent Python 12 then?

For the time being we should probably not mark packages broken for 3.12 to eagerly.

**Review by CorbanR — APPROVED** [review]

File: nixos/doc/manual/release-notes/rl-2311.section.md (hunk 7 above)
Comment (SuperSandro2000) [reply nixos/doc/manual/release-notes/rl-2311.section.md]: I am not sure what you want to restrict.

**Comment by CorbanR** [comment]
@gaelreyrol I think you may need to fix some branch conflicts.

**Comment by gaelreyrol** [comment]
> @gaelreyrol I think you may need to fix some branch conflicts.

Sorry for the delay, I rebased my PR to clean the conflicts.

File: pkgs/development/libraries/libpulsar/default.nix [hunk 9]
```diff
@@ -37,48 +33,36 @@ let
...
-  version = "2.10.2";
+  version = "3.3.0";
 
-  src = fetchurl {
-    hash = "sha256-IONnsSDbnX2qz+Xya0taHYSViTOiRI36AfcxmY3dNpo=";
-    url = "mirror://apache/pulsar/pulsar-${version}/apache-pulsar-${version}-src.tar.gz";
+  src = fetchFromGitHub {
+    owner = "apache";
+    repo = "pulsar-client-cpp";
+    rev = "v${finalAttrs.version}";
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/libpulsar/default.nix:50]:
```suggestion
    rev = "v${version}";
```
that won't work unless hash is overwritten which easily hides bugs

File: pkgs/development/libraries/libpulsar/default.nix [hunk 10]
```diff
@@ -92,11 +76,11 @@ stdenv.mkDerivation rec {
   '';
 
   meta = with lib; {
-    homepage = "https://pulsar.apache.org/docs/en/client-libraries-cpp";
+    homepage = "https://pulsar.apache.org/docs/3.0.x/client-libraries-cpp/";
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/libpulsar/default.nix:79]:
```suggestion
    homepage = "https://pulsar.apache.org/docs/3.1.x/client-libraries-cpp/";
```
why do the version numbers not match up? strange

File: pkgs/development/python-modules/pulsar/default.nix [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+    make install
+    cd ..
+  '';
+
+  propagatedBuildInputs = [
+    certifi
+  ];
+
+  # Requires to setup a cluster
+  doCheck = false;
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/pulsar/default.nix:47]: We should add a pythonImportsCheck

File: pkgs/tools/misc/otel-trace/default.nix [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+  src = fetchFromGitHub {
+    owner = "Pondidum";
+    repo = "Trace";
+    rev = version;
+    hash = "sha256-mhPEGCu9nikRetiTO0fUj2BkEX510CGbC2hn/mXyItQ=";
+  };
+
+  vendorHash = "sha256-unZMA1S9rYRlAitYmVIrNMTI1Cg0oevk2HWSoc6wLFs=";
+
+  preBuild = "go generate ./...";
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/otel-trace/default.nix:22]:
```suggestion
  preBuild = ''
    go generate ./...
  '';
```

File: pkgs/tools/misc/otel-trace/default.nix [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+
+  ldflags = [
+    "-s"
+    "-w"
+    # Requires raw commit hash because version is picked up from GitCommit[0:7] which produce an out of bounds slice range runtime error.
+    "-X trace/version.GitCommit=\"1bb1b9a2f55337e8ecc567d13e905e1ed46267c2\""
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/otel-trace/default.nix:28]: We could also mock this with 32 e's or so

**Review by SuperSandro2000 — APPROVED** [review]

**Review by CorbanR — APPROVED** [review]

**Review by CorbanR — APPROVED** [review]

**Review by SuperSandro2000 — APPROVED** [review]

File: pkgs/development/python-modules/pulsar/default.nix [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+
+  pythonImportsCheck = [];
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/pulsar/default.nix:49]: We also need to add an entry here, otherwise the build is not doing much.

**Comment by alippai** [comment]
Does this add `pulsar-admin` as well? Or was it already included?

**Comment by gaelreyrol** [comment]
> Does this add `pulsar-admin` as well? Or was it already included?

No it's not, this PR is only about the CPP and Python libraries

File: pkgs/development/libraries/libpulsar/default.nix [hunk 15]
```diff
@@ -37,48 +35,39 @@ let
...
-  defaultOptionals = [ boost protobuf ]
-    ++ lib.optional python3Support python3
+  defaultOptionals = [ protobuf ]
     ++ lib.optional snappySupport snappy.dev
     ++ lib.optional zlibSupport zlib
     ++ lib.optional zstdSupport zstd
-    ++ lib.optional log4cxxSupport log4cxx;
+    ++ lib.optional log4cxxSupport log4cxx
+    ++ lib.optional asioSupport asio
+    ++ lib.optional (asioSupport == false) boost180;
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/libpulsar/default.nix:44]:
```suggestion
    ++ lib.optional (!asioSupport) boost180;
```

**Review by CorbanR — APPROVED** [review]

**Pull request merged**